controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -944,118 +944,158 @@ private boolean isSubscriberToNonSubscriberTransition(final ReaderGroupConfigRec
-        // if the lowerbound on subscribers streamcuts satisfies the policy size bound, then return it. 
-        // else return the stream cut that satisfies maximum bound on size. 
         // 1. if lowerbound.size < max and lowerbound.size > min truncate at lowerbound
-        // 2. if lowerbound.size < min, truncate at max irrespective of if lowerbound overlaps with max or not. 
+        // 2. if lowerbound.size < min, truncate at streamcut behind lowerbound that satisfies the policy. (we look at max and below).
         // 3. if lowerbound.size > max, truncate at max
         long currentSize = retentionSet.getLatest().getRecordingSize();
+        Map.Entry<StreamCutReferenceRecord, StreamCutReferenceRecord> limits = getBoundStreamCuts(policy, retentionSet,
+                x -> currentSize - x.getRecordingSize());
 
         // if lowerbound is empty simply return min
         if (lowerBound == null || lowerBound.isEmpty()) {
-            return retentionSet
-                    .getRetentionRecords().stream()
-                    .filter(x -> currentSize - x.getRecordingSize() > policy.getRetentionParam())
-                    .max(Comparator.comparingLong(StreamCutReferenceRecord::getRecordingTime))
+            return Optional.ofNullable(limits.getValue())
                     .map(x -> streamMetadataStore.getStreamCutRecord(scope, stream, x, context, executor).thenApply(StreamCutRecord::getStreamCut))
                     .orElse(CompletableFuture.completedFuture(null));
         }
 
         return streamMetadataStore.getSizeTillStreamCut(scope, stream, lowerBound, Optional.empty(), context, executor)
-                .thenCompose(sizeTill -> {
-                    long retainedSize = currentSize - sizeTill;
-                    Supplier<Optional<StreamCutReferenceRecord>> maxBound = () -> retentionSet
-                    .getRetentionRecords().stream()
-                    .filter(x -> currentSize - x.getRecordingSize() > policy.getRetentionMax())
-                    .max(Comparator.comparingLong(StreamCutReferenceRecord::getRecordingTime));
-                    // if retainedSize is less than min size then do not truncate the stream.
-                    if (retainedSize < policy.getRetentionParam()) {
-                        // if retainedSize is less than min size then truncate the stream at max bound.
-                        return maxBound.get().map(x -> streamMetadataStore.getStreamCutRecord(scope, stream, x, context, executor)
-                        .thenApply(StreamCutRecord::getStreamCut))
-                        .orElse(CompletableFuture.completedFuture(null));
-                        } else {
+                .thenCompose(sizeTillLB -> {
+                    long retainedSizeLB = currentSize - sizeTillLB;
+                    // if retainedSize is less than min size then we need to truncate at min or the most recent streamcut
+                    // strictly behind lowerbound. we actually choose the max
```

Comment:
Didn't get this part of the comment: 
`the most recent streamcut strictly behind lowerbound. we actually choose the max`

I thought it should be truncated at a stream-cut that satisfies min

Reply:
we truncate at min if it doesnt overlap with SLB.. if it does, we choose the first streamcut from retentionset strictly behind SLB.
the comment about max is not correct.. removed it.